File: python/src/benchmark/data_v2/protobuf_bridge.py
"""Map v2 dataclasses ↔ generated benchmark.v2 protobuf messages."""

from __future__ import annotations

import os
import sys
from typing import Any, List, Type, Union

from .models import Document, DocumentItem, DocumentMeta, Event, EventAttr, Message, Strings, Telemetry

# python/generated is on path when running as package (same as v1 protobuf).
_gen_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "generated"))
if _gen_root not in sys.path:
    sys.path.insert(0, _gen_root)
_py_root = os.path.abspath(os.path.join(_gen_root, ".."))
if _py_root not in sys.path:
    sys.path.insert(0, _py_root)

try:
    from generated.v2 import benchmark_v2_pb2 as pb2  # type: ignore
except ImportError:
    try:
        from v2 import benchmark_v2_pb2 as pb2  # type: ignore
    except ImportError:
        pb2 = None  # type: ignore


def available() -> bool:
    return pb2 is not None


def message_class_for(type_id: str, batch: bool) -> Type[Any] | None:
    if pb2 is None:
        return None
    table = {
        ("message", False): pb2.Message,
        ("message", True): pb2.BatchMessage,
        ("document", False): pb2.Document,
        ("document", True): pb2.BatchDocument,
        ("telemetry", False): pb2.Telemetry,
        ("telemetry", True): pb2.BatchTelemetry,
        ("strings", False): pb2.Strings,
        ("strings", True): pb2.BatchStrings,
        ("event", False): pb2.Event,
        ("event", True): pb2.BatchEvent,
    }
    return table.get((type_id, batch))


def to_pb(obj: Any) -> Any:
    if pb2 is None:
        raise RuntimeError("v2 protobuf stubs not available")
    if isinstance(obj, list):
        if not obj:
            raise ValueError("empty batch")
        batch = _batch_for_item(obj[0])
        for item in obj:
            batch.items.append(to_pb(item))
        return batch
    if isinstance(obj, Message):
        m = pb2.Message()
        m.f_bool = obj.f_bool
        m.f_int32 = obj.f_int32
        m.f_int64 = obj.f_int64
        m.f_float64 = obj.f_float64
        m.f_string = obj.f_string
        m.f_bool_2 = obj.f_bool_2
        m.f_int32_2 = obj.f_int32_2
        m.f_string_2 = obj.f_string_2
        return m
    if isinstance(obj, Document):
        d = pb2.Document()
        d.id = obj.id
        d.status = obj.status
        d.meta.region = obj.meta.region
        d.meta.version = obj.meta.version
        for it in obj.items:
            row = d.items.add()
            row.sku = it.sku
            row.qty = it.qty
            row.price_minor = it.price_minor
        return d
    if isinstance(obj, Telemetry):
        t = pb2.Telemetry()
        t.source = obj.source
        t.ts = obj.ts
        t.tags.extend(obj.tags)
        t.values.extend(obj.values)
        return t
    if isinstance(obj, Strings):
        s = pb2.Strings()
        s.items.extend(obj.items)
        return s
    if isinstance(obj, Event):
        e = pb2.Event()
        e.event_id = obj.event_id
        e.event_type = obj.event_type
        e.occurred_at = obj.occurred_at
        e.producer = obj.producer
        for a in obj.attrs:
            row = e.attrs.add()
            row.key = a.key
            row.value = a.value
        return e
    raise TypeError(f"unsupported v2 type for protobuf: {type(obj)}")


def _batch_for_item(item: Any) -> Any:
    if isinstance(item, Message):
        return pb2.BatchMessage()
    if isinstance(item, Document):
        return pb2.BatchDocument()
    if isinstance(item, Telemetry):
        return pb2.BatchTelemetry()
    if isinstance(item, Strings):
        return pb2.BatchStrings()
    if isinstance(item, Event):
        return pb2.BatchEvent()
    raise TypeError(type(item))

File: swift/Sources/SerializerBenchmarkCore/Serializers/ProtobufBridge.swift
import Foundation
import SwiftProtobuf

/// Domain ↔ SwiftProtobuf conversion (type-aware). Used only in prepare / post-deser fidelity path.
enum ProtobufBridge {
    static func toProtobuf(_ fixture: Fixture) throws -> any SwiftProtobuf.Message {
        if fixture.instanceCount > 1 {
            return try toBatch(fixture)
        }
        switch fixture.name {
        case "message":
            return messagePB(fixture.value as! Message)
        case "document":
            return documentPB(fixture.value as! Document)
        case "telemetry":
            return telemetryPB(fixture.value as! Telemetry)
        case "strings":
            return stringsPB(fixture.value as! Strings)
        case "event":
            return eventPB(fixture.value as! Event)
        default:
            throw BenchError.unknownType(fixture.name)
        }
    }

    static func toDomain(_ msg: any SwiftProtobuf.Message, fixture: Fixture) throws -> Any {
        if fixture.instanceCount > 1 {
            return try fromBatch(msg, fixture: fixture)
        }
        switch fixture.name {
        case "message":
            return messageDomain(msg as! Benchmark_V2_Message)
        case "document":
            return documentDomain(msg as! Benchmark_V2_Document)
        case "telemetry":
            return telemetryDomain(msg as! Benchmark_V2_Telemetry)
        case "strings":
            return stringsDomain(msg as! Benchmark_V2_Strings)
        case "event":
            return eventDomain(msg as! Benchmark_V2_Event)
        default:
            throw BenchError.unknownType(fixture.name)
        }
    }

    private static func toBatch(_ fixture: Fixture) throws -> any SwiftProtobuf.Message {
        switch fixture.name {
        case "message":
            var b = Benchmark_V2_BatchMessage()
            b.items = (fixture.value as! [Message]).map(messagePB)
            return b
        case "document":
            var b = Benchmark_V2_BatchDocument()
            b.items = (fixture.value as! [Document]).map(documentPB)
            return b
        case "telemetry":
            var b = Benchmark_V2_BatchTelemetry()
            b.items = (fixture.value as! [Telemetry]).map(telemetryPB)
            return b
        case "strings":
            var b = Benchmark_V2_BatchStrings()
            b.items = (fixture.value as! [Strings]).map(stringsPB)
            return b
        case "event":
            var b = Benchmark_V2_BatchEvent()
            b.items = (fixture.value as! [Event]).map(eventPB)
            return b
        default:
            throw BenchError.unknownType(fixture.name)
        }
    }

    private static func fromBatch(_ msg: any SwiftProtobuf.Message, fixture: Fixture) throws -> Any {
        switch fixture.name {
        case "message":
            return (msg as! Benchmark_V2_BatchMessage).items.map(messageDomain)
        case "document":
            return (msg as! Benchmark_V2_BatchDocument).items.map(documentDomain)
        case "telemetry":
            return (msg as! Benchmark_V2_BatchTelemetry).items.map(telemetryDomain)
        case "strings":
            return (msg as! Benchmark_V2_BatchStrings).items.map(stringsDomain)
        case "event":
            return (msg as! Benchmark_V2_BatchEvent).items.map(eventDomain)
        default:
            throw BenchError.unknownType(fixture.name)
        }
    }

    static func messagePB(_ m: Message) -> Benchmark_V2_Message {
        var p = Benchmark_V2_Message()
        p.fBool = m.f_bool
        p.fInt32 = m.f_int32
        p.fInt64 = m.f_int64
        p.fFloat64 = m.f_float64
        p.fString = m.f_string
        p.fBool2 = m.f_bool_2
        p.fInt322 = m.f_int32_2
        p.fString2 = m.f_string_2
        return p
    }
    static func messageDomain(_ p: Benchmark_V2_Message) -> Message {
        Message(
            f_bool: p.fBool, f_int32: p.fInt32, f_int64: p.fInt64, f_float64: p.fFloat64,
            f_string: p.fString, f_bool_2: p.fBool2, f_int32_2: p.fInt322, f_string_2: p.fString2
        )
    }

    static func documentPB(_ d: Document) -> Benchmark_V2_Document {
        var p = Benchmark_V2_Document()
        p.id = d.id
        p.status = d.status
        var meta = Benchmark_V2_DocumentMeta()
        meta.region = d.meta.region
        meta.version = d.meta.version
        p.meta = meta
        p.items = d.items.map { it in
            var x = Benchmark_V2_DocumentItem()
            x.sku = it.sku; x.qty = it.qty; x.priceMinor = it.price_minor
            return x
        }
        return p
    }
    static func documentDomain(_ p: Benchmark_V2_Document) -> Document {
        Document(
            id: p.id, status: p.status,
            meta: DocumentMeta(region: p.meta.region, version: p.meta.version),
            items: p.items.map { DocumentItem(sku: $0.sku, qty: $0.qty, price_minor: $0.priceMinor) }
        )
    }

    static func telemetryPB(_ t: Telemetry) -> Benchmark_V2_Telemetry {
        var p = Benchmark_V2_Telemetry()
        p.source = t.source; p.ts = t.ts; p.tags = t.tags; p.values = t.values
        return p
    }
    static func telemetryDomain(_ p: Benchmark_V2_Telemetry) -> Telemetry {
        Telemetry(source: p.source, ts: p.ts, tags: p.tags, values: p.values)
    }

    static func stringsPB(_ s: Strings) -> Benchmark_V2_Strings {
        var p = Benchmark_V2_Strings(); p.items = s.items; return p
    }
    static func stringsDomain(_ p: Benchmark_V2_Strings) -> Strings {
        Strings(items: p.items)
    }

    static func eventPB(_ e: Event) -> Benchmark_V2_Event {
        var p = Benchmark_V2_Event()
        p.eventID = e.event_id
        p.eventType = e.event_type
        p.occurredAt = e.occurred_at
        p.producer = e.producer
        p.attrs = e.attrs.map { a in
            var x = Benchmark_V2_EventAttr(); x.key = a.key; x.value = a.value; return x
        }
        return p
    }
    static func eventDomain(_ p: Benchmark_V2_Event) -> Event {
        Event(
            event_id: p.eventID, event_type: p.eventType, occurred_at: p.occurredAt,
            producer: p.producer,
            attrs: p.attrs.map { EventAttr(key: $0.key, value: $0.value) }
        )
    }
}
